### PR TITLE
Fix range contains

### PR DESCRIPTION
**Pull request by JoaquinTrinanes**
# Description

This PR changes the range contains logic to take the step into account. 

```nushell
# before
2 in 1..3.. # true

# now
2 in 1..3.. # false
```

---

I encountered another issue while adding tests. Due to floating point precision, `2.1 in 1..1.1..3` will return `false`. The floating point error is even bigger than `f64::EPSILON` (`0.09999999999999876` vs `2.220446049250313e-16`). This issue disappears with bigger numbers.

I tried a different algorithm (checking if the estimated number of steps is close enough to any integer) but the results are still pretty bad:

```rust
let n_steps = (value - self.start) / self.step; // 14.999999999999988
(n_steps - n_steps.round()).abs() < f64::EPSILON // returns false
```

Maybe it can be shipped like this, the REPL already has floating point errors (`1.1 - 1` returns `0.10000000000000009`). Or maybe there's a way to fix this that I didn't think of. I'm open to ideas! But in any case performing this kind of checks on a range of floats seems more niche than doing it on a range of ints.

# User-Facing Changes

Code that depended on this behavior to check if a number is between `start` and `end` will potentially return a different value.

# Tests + Formatting

# After Submitting

### CODE REVIEW

File: crates/nu-protocol/src/value/range.rs [hunk 1]
```diff
@@ -95,20 +95,28 @@ mod int_range {
 
         pub fn contains(&self, value: i64) -> bool {
             if self.step < 0 {
-                value <= self.start
-                    && match self.end {
-                        Bound::Included(end) => value >= end,
-                        Bound::Excluded(end) => value > end,
-                        Bound::Unbounded => true,
-                    }
+                // Decreasing range
```
Comment (JoaquinTrinanes) [inline crates/nu-protocol/src/value/range.rs:98]: I could probably normalize the range (set start to the minimum value, end to the highest) and just apply this logic once

File: crates/nu-protocol/src/value/range.rs [hunk 2]
```diff
@@ -95,20 +95,28 @@ mod int_range {
...
             } else {
-                self.start <= value
-                    && match self.end {
-                        Bound::Included(end) => value <= end,
-                        Bound::Excluded(end) => value < end,
-                        Bound::Unbounded => true,
-                    }
+                // Increasing range
+                if value < self.start {
+                    dbg!(value, self.start, value > self.start);
```
Comment (1Kinoti) [inline crates/nu-protocol/src/value/range.rs:110]: i think you should probably remove this `dbg!()`
```suggestion

```

Comment (JoaquinTrinanes) [reply crates/nu-protocol/src/value/range.rs]: Definitely, I missed that one when cleaning up 😅

**Comment by NotTheDr01ds** [comment]
> Maybe it can be shipped like this, the REPL already has floating point errors (`1.1 - 1` returns `0.10000000000000009`). Or maybe there's a way to fix this that I didn't think of. I'm open to ideas!

I notice that `1.1 == 0.10000000000000009` returns `true`.  Is there anything in that float comparison code that can help here?  I haven't looked yet, and I'm guessing "no", but perhaps worth a shot.

> But in any case performing this kind of checks on a range of floats seems more niche than doing it on a range of ints.

Personally, I agree

**Comment by NotTheDr01ds** [comment]
Btw, I've tested the PR and it works as advertised ;-)

**Comment by JoaquinTrinanes** [comment]
> I notice that 1.1 == 0.10000000000000009 returns true. Is there anything in that float comparison code that can help here? I haven't looked yet, and I'm guessing "no", but perhaps worth a shot.

I had already tried using the method the equals operator uses (`precision = f64::EPSILON.max(value.abs() * f64::EPSILON)`) without much success (makes sense, the values we're working with are close to 1). But after debugging around it turns out the error comes from the range creation itself.

The constructor takes `start`, `next` and `end`. And then calculates the step as `next - start`, which in the failing example translates to `1.1 - 1`. Other than resorting to a big decimal library or limiting the precision I can't think of anything that would fix the issue 🥲

**Comment by fdncred** [comment]
Thanks

**Pull request merged**